Utils/Dataflow/095_datasetInfoAMI/amiDatasets.py

```diff
@@ -50,12 +50,15 @@ def main(argv):
 
     stage.configure(argv)
     stage.process = process
-    if stage.ARGS.userkey and stage.ARGS.usercert:
+    try:
         init_ami_client(stage.ARGS.userkey, stage.ARGS.usercert)
-    exit_code = stage.run()
+    except Exception:
+        exit_code = 1
+    else:
+        exit_code = stage.run()
```

Comment:
While the change, indeed, breaks `--skip` usage and should be fixed - thanks for pointing this out - I'm not sure that `if` should be left in place as-is.

Commits related to the issue:

1. dc13dc7 - here the `if` is introduced, precisely for working with `--skip`.
2. ce05ca8 - this commit unintentionally introduces the problem: message 'Result' for each dataset starting with the second when no credentials were supplied. This happens because calling `get_ami_client()` in `amiPhysValues()` leads to exception: before the commit the call is performed normally and the exception stops the program execution, while after the commit it is caught by `try-except` and the execution continues.
3. ce3841d - here the problem is solved by always trying to perform the initialization, regardless of (un)supplied credentials, and stopping the execution if something is wrong.

Now, while the initial code (one in `master`) works correctly - it tries to process one dataset and stops due to missing credentials - it does so due to the fact that this first processing throws an unhandled exception. While such logic can be left as-is in some cases, I believe that this one is different - after all, one of the "symptoms" that led to this whole PR was the code talking about "no data in AMI" for each dataset when credentials were missing.

Therefore, I believe that the logic "if certificate and key are supplied, initialize the client" should be replaced with something like "if `--skip` is absent, initialize the client". This way, missing credentials will be properly caught and reported once, while `--skip` will function properly. Will look into it further to determine how the history should be rewritten in #312 to be easier to understand.

Reply:
@Evildoor,
> Now, while the initial code (one in master) works correctly - it tries to process one dataset and stops due to missing credentials - it does so due to the fact that this first processing throws an unhandled exception. 

1. I wouldn't call it an ***unhandled*** exception. Technically, "unhandled" is the one that reaches the outermost context of the executed program without being caught. It looks like this:
    ```
    (INFO) Start program execution.
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    Exception: This is an unhandled exception!
    ```
    But there are situations when exceptions indicate that program should be stopped, right? Then it may look like this:
    ```
    (INFO) Start program execution.
    (ERROR) Some error occurred!
    (DEBUG) Traceback (most recent call last):
    (DEBUG)   File "<stdin>", line 1, in <module>
    (DEBUG) Exception: Some error occured!
    (INFO) Stop program execution.
    ```
    -- the exception was caught, analyzed, and since it is not possible to go on with execution -- user was provided with debug information about the situation, and the program was gracefully stopped.

    Here's what we have in `master`:
    ```
    2020-01-17 11:23:22 (INFO) (ProcessorStage) Starting stage execution.
    (ERROR) Login authentication mode is not supported. Please provide user certificate or create proxy.
    2020-01-17 11:23:22 (ERROR) (ProcessorStage) Failed to initialise AMI client: certificate not provided or not found.
    2020-01-17 11:23:22 (DEBUG) (ProcessorStage) Traceback (most recent call last):
    (==)   File "/home/dkb/dkb-dev.git/Utils/Dataflow/095_datasetInfoAMI/../pyDKB/dataflow/stage/ProcessorStage.py", line 231, in run
    (==)     if msg and process(self, msg):
    (==)   File "./amiDatasets.py", line 107, in process
    (==)     amiPhysValues(data)
    (==)   File "./amiDatasets.py", line 123, in amiPhysValues
    (==)     ami_client = get_ami_client()
    (==)   File "./amiDatasets.py", line 89, in get_ami_client
    (==)     init_ami_client()
    (==)   File "./amiDatasets.py", line 82, in init_ami_client
    (==)     raise DataflowException("Failed to initialise AMI client: certificate "
    (==) DataflowException: Failed to initialise AMI client: certificate not provided or not found.
    2020-01-17 11:23:22 (INFO) (ProcessorStage) Stopping stage.
    ```
    Looks pretty handled to me, m?

2. It is "unhandled" from the `process()` point of view -- but it only means one thing: it should be properly handled. And, no matter what -- I insist that "proper handling" in this situation would be to gracefully stop the program execution. "Default" handling policy was "not to handle" == "propagate all exceptions" and it worked properly. When `try/except clause` was added, it changed. And it is exactly what I wanted you to analyze: what has changed, why it lead to the new bug, why previous version worked properly, and -- the most important part, -- are there other but "`--skip` case" situations, when it may possibly lead to an unexpected behaviour?

3. > ce3841d - here the problem is solved by always trying to perform the initialization, regardless of (un)supplied credentials, and stopping the execution if something is wrong.

    ...and introduces another problem: `--skip` option does not work properly.

    What I am saying, is that in situations like this: `code works` -> `change X introduces bug A` -> `fixing bug A introduces bug B` -- the "bug B" point is a good moment to stop making changes and take a closer look at the code. Because this situation means that author of the changes does not fully understand the logic of the code, and next fix may as well break something else... and who can guarantee that new issue will be detected immediately? I mean, if we had unit tests -- that's easier, just go-go-go, fix-fix-fix till all the tests say OK (although it would also be a good idea to stop for a moment and think about what's going on). But since we don't have them, there's a good chance that new issue stays undetected till someone steps on it -- and hopefully it will happen a controlled test environment, not in "production" version where it may do some damage.

4. > While such logic (<stop execution due to an unhandled exception>) can be left as-is in some cases, I believe that this one is different - after all, one of the "symptoms" that led to this whole PR was the code talking about "no data in AMI" for each dataset when credentials were missing.

    I am sorry, but (leaving aside that the logic is a bit different, see p.1), I don't see how from "program incorrectly handles situation with missed credentials" follows "program shouldn't stop execution due to an exception thrown during client initialization".

5. > Therefore, I believe that the logic "if certificate and key are supplied, initialize the client" should be replaced with something like "if --skip is absent, initialize the client". This way, missing credentials will be properly caught and reported once, while `--skip` will function properly.

    Although the "therefore"-preceding part sounds very doubtful to me, what you suggest worth trying: for now we seem to have only one use-case when we don't need to initialize the client, so it's better to fail early if the client can not be initialized. However, I want you to understand: ***it does not fix the bug -- it eliminates a detected symptom***.

---

Summary:
* the more code is changed, the more the probability of inviting new bugs;
* => to fix some bug, it's better to change less than more -- and therefore it's better to change what has introduced the bug, not the rest of the code;
* to change the way of allowing `--skip` will not fix the issue, only make it less visible (*maybe* removing the only possible effect ever, but I doubt it);
* but the way `--skip`-allowing logic was implemented is not good indeed and worth improving *(which is one of the reasons why do a code review: to improve things that worth improving when it's "work in progress", not "discussed, merged and forgotten for good" ;) )*.